src/oic/oic/provider.py

```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
+
+        if "backchannel_logout_uri" in _cdb[client_id]:
+            _sub = _sso_db[sid]["sub"]
+            _spec = self.do_back_channel_logout(_cdb[client_id], _sub, sid)
+            if _spec:
+                res["blu"] = {client_id: _spec}
+        elif "frontchannel_logout_uri" in _cdb[client_id]:
+            # Construct an IFrame
+            _iframe = do_front_channel_logout_iframe(_cdb[client_id], self.name, sid)
+            if _iframe:
+                res["flu"] = {client_id: _iframe}
+
+        # Clean out sessions
+        self.clean_sessions([sid])
+
+        return res
+
+    def end_session_endpoint(
+        self, request: str = "", cookie: Union[str, SimpleCookie, None] = None, **kwargs
+    ) -> Response:
+        """
+        Handle a RP initiated Logout request.
+
+        :param request: The logout request
+        :param cookie:
+        :param kwargs:
+        :return: Returns a dictionary with one key 'sjwt' and the value
+            being a signed JWT token with session information.
+        """
+        esr = EndSessionRequest().from_urlencoded(request)
+
+        logger.debug("End session request: {}", sanitize(esr.to_dict()))
+
+        if self.events:
+            self.events.store("protocol request", esr)
+
+        # 2 ways of find out client ID and user. Either through a cookie
+        # or using the id_token_hint. If I get information from both make sure they match
+        _, client_id, uid = self._get_uid_from_cookie(cookie)
+
+        sid = ""
+
+        if "id_token_hint" in esr:
+            id_token_hint = IdToken().from_jwt(
+                esr["id_token_hint"], keyjar=self.keyjar, verify=True
+            )
+            far_away = 86400 * 30  # 30 days
+
+            if client_id:
+                args = {"client_id": client_id}
+            else:
+                args = {}
+
+            try:
+                id_token_hint.verify(
+                    iss=self.baseurl, skew=far_away, nonce_storage_time=far_away, **args
+                )
+            except (VerificationError, NotForMe) as err:
+                logger.warning("Verification error on id_token_hint: {}".format(err))
+                return error_response("invalid_request", "Bad Id Token hint")
+
+            sub = id_token_hint["sub"]
+
+            if uid:
+                # verify that 'sub' are bound to 'uid'
+                if self.sdb.get_uid_by_sub(sub) != uid:
+                    return error_response("invalid_request", "Wrong user")
+            else:
+                uid = self.sdb.get_uid_by_sub(sub)
+
+            if not client_id:
+                if len(id_token_hint["aud"]) == 1:
+                    client_id = id_token_hint["aud"][0]
+                else:
+                    client_id = id_token_hint["azp"]
+
+            sids = session_get(self.sdb, "sub", sub)
+
+            for sid in sids:
+                if self.sdb[sid]["client_id"] == client_id:
+                    break
+
+        if not client_id:
```

Comment:
Again, test for `None` should be explicit.

Reply:
OK